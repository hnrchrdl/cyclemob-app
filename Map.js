import React from 'react';
import { Dimensions, StyleSheet } from 'react-native';
import MapView, { MAP_TYPES, UrlTile } from 'react-native-maps';

const { width, height } = Dimensions.get('window');

const ASPECT_RATIO = width / height;
const LATITUDE_DELTA = 0.0922;
const LONGITUDE_DELTA = LATITUDE_DELTA * ASPECT_RATIO;
const TILE_URLS = {
  ocm:
    'https://tile.thunderforest.com/cycle/{z}/{x}/{y}.png?apikey=61444b383bbc468dbd554f7257efd5f3'
};

class Map extends React.Component {
  constructor(props, context) {
    super(props, context);
    this.state = {
      iniitalRegion: null
    };
  }

  componentDidMount() {
    navigator.geolocation.getCurrentPosition(
      pos => {
        console.log(pos);
        const { latitude, longitude } = pos.coords;
        if (!this.state.initialRegion) {
          this.setState({
            initialRegion: {
              latitude,
              longitude,
              latitudeDelta: LATITUDE_DELTA,
              longitudeDelta: LONGITUDE_DELTA
            }
          });
        }
      },
      err => {
        console.err(err);
      },
      { enableHighAccuracy: true, timeout: 20000, maximumAge: 1000 }
    );
  }

  handleMapLongPress = e => {
    console.log('long press', e);
  };

  render() {
    const { initialRegion } = this.state;
    if (!initialRegion) {
      return null;
    }
    return (
      <MapView
        style={styles.map}
        mapType={MAP_TYPES.NONE}
        initialRegion={initialRegion}
        showsUserLocation
        followsUserLocation={this.state.followUser}
        onLongPress={this.handleMapLongPress}
      >
        <UrlTile urlTemplate={TILE_URLS.ocm} zIndex={-1} />
      </MapView>
    );
  }
}

const styles = StyleSheet.create({
  map: {
    position: 'absolute',
    top: 0,
    left: 0,
    right: 0,
    bottom: 0
  }
});

export default Map;
