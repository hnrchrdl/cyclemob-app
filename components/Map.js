import React from 'react';
import { Dimensions, StyleSheet } from 'react-native';
import PropTypes from 'prop-types';
import MapView, {
  MAP_TYPES,
  UrlTile,
  Marker,
  Polyline
} from 'react-native-maps';
import { thunderforest_api_key } from '../env';

const { width, height } = Dimensions.get('window');
const ASPECT_RATIO = width / height;
const TILE_URLS = {
  ocm: `https://tile.thunderforest.com/cycle/{z}/{x}/{y}.png?apikey=${thunderforest_api_key}`
};

function getRegionFromPositionAndZoom(position, zoomFactor) {
  const latitude = position.coords.latitude;
  const longitude = position.coords.longitude;
  const latitudeDelta = 0.004 * zoomFactor;
  const longitudeDelta = latitudeDelta * ASPECT_RATIO;
  const region = {
    latitude,
    longitude,
    latitudeDelta,
    longitudeDelta
  };
  return region;
}

class Map extends React.PureComponent {
  map = null;

  componentDidUpdate(prevProps) {
    if (this.map) {
      if (
        (this.props.route &&
          JSON.stringify(prevProps.route) !==
            JSON.stringify(this.props.route)) ||
        (this.props.marker &&
          JSON.stringify(prevProps.marker) !==
            JSON.stringify(this.props.marker))
      ) {
        // Route or Marker changed.
        const coords = [
          ...(this.props.position
            ? [
              {
                latitude: this.props.position.coords.latitude,
                longitude: this.props.position.coords.longitude
              }
            ]
            : []),
          ...(this.props.route ? this.props.route : []),
          ...(this.props.marker ? this.props.marker.map(m => m.coordinate) : [])
        ];
        if (coords.length > 1) {
          this.map.fitToCoordinates(coords, {
            edgePadding: { top: 100, left: 50, right: 50, bottom: 200 },
            animated: true
          });
        }
      } else if (this.props.position && this.props.followPosition && this.map) {
        // Position or FollowPosition changed.
        this.map.animateToCoordinate(this.props.position.coords);
      }
      if (this.props.zoom && prevProps.zoom !== this.props.zoom) {
        // Zoom changed.
        const region = getRegionFromPositionAndZoom(
          this.props.position,
          this.props.zoom
        );
        this.map.animateToRegion(region);
      }
    }
  }

  onMapLongPress = () => {
    if (this.props.onMapLongPressed) {
      this.props.onMapLongPressed();
    }
  };

  onMarkerPressed = e => {
    if (this.props.onMarkerPressed) {
      this.props.onMarkerPressed(e);
    }
  };

  isActive = marker => marker.id === this.props.activeMarkerId;

  render() {
    if (!this.props.position) {
      return null;
    }

    const { marker, route, position } = this.props;
    const region = getRegionFromPositionAndZoom(position, 5);

    return (
      <MapView
        ref={el => (this.map = el)}
        style={styles.map}
        mapType={MAP_TYPES.NONE}
        initialRegion={region}
        showsUserLocation={true}
        onLongPress={this.onMapLongPress}
      >
        <UrlTile urlTemplate={TILE_URLS.ocm} zIndex={-1} />
        {marker &&
          marker.map(_marker => (
            <Marker
              key={_marker.id}
              coordinate={_marker.coordinate}
              pinColor={this.isActive(_marker) ? 'tomato' : 'gold'}
              onPress={() => {
                this.onMarkerPressed(_marker);
              }}
            />
          ))}
        {route && (
          <Polyline
            coordinates={route}
            strokeWidth={5}
            strokeColor={'rgba(255,99,71,0.7)'}
          />
        )}
      </MapView>
    );
  }
}
Map.propTypes = {
  position: PropTypes.shape({
    coords: PropTypes.shape({
      latitude: PropTypes.number,
      longitude: PropTypes.number
    })
  }),
  followPosition: PropTypes.bool.isRequired,
  marker: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.string,
      coordinate: PropTypes.shape({
        latitude: PropTypes.number,
        longitude: PropTypes.number
      })
    })
  ).isRequired,
  activeMarkerId: PropTypes.string,
  onMarkerPressed: PropTypes.func,
  route: PropTypes.arrayOf(
    PropTypes.shape({
      latitude: PropTypes.number,
      longitude: PropTypes.number
    })
  ),
  onMapLongPressed: PropTypes.func,
  zoom: PropTypes.number // 0 for all zoomed in, higher numbers for wider map zoom
};

const styles = StyleSheet.create({
  map: {
    position: 'absolute',
    top: 0,
    left: 0,
    right: 0,
    bottom: 0
  }
});

export default Map;
