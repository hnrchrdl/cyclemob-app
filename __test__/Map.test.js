import React from 'react';
import Map from '../components/Map';

import { shallow } from 'enzyme';

jest.mock('react-native-maps', () => {
  const React = require.requireActual('react');
  const MapView = require.requireActual('react-native-maps');

  class MockCallout extends React.Component {
    render() {
      // eslint-disable-next-line react/prop-types
      return React.createElement('Callout', this.props, this.props.children);
    }
  }

  class MockMarker extends React.Component {
    render() {
      // eslint-disable-next-line react/prop-types
      return React.createElement('Marker', this.props, this.props.children);
    }
  }

  class MockMapView extends React.Component {
    render() {
      // eslint-disable-next-line react/prop-types
      return React.createElement('MapView', this.props, this.props.children);
    }
  }

  class MockUrlTile extends React.Component {
    render() {
      // eslint-disable-next-line react/prop-types
      return React.createElement('UrlTile', this.props, this.props.children);
    }
  }

  class MockPolyline extends React.Component {
    render() {
      // eslint-disable-next-line react/prop-types
      return React.createElement('Polyline', this.props, this.props.children);
    }
  }

  MockCallout.propTypes = MapView.Callout.propTypes;
  MockMarker.propTypes = MapView.Marker.propTypes;
  MockMapView.propTypes = MapView.propTypes;
  MockMapView.Marker = MockMarker;
  MockMapView.Callout = MockCallout;
  MockUrlTile.propTypes = MapView.UrlTile.propTypes;
  MockMapView.UrlTile = MockUrlTile;
  MockMapView.Polyline = MockPolyline;
  MockPolyline.propTypes = MapView.Polyline.propTypes;
  MockMapView.MAP_TYPES = { NONE: 'None' };
  return MockMapView;
});

const props = {
  position: {
    coords: {
      latitude: 50,
      longitude: 10
    }
  },
  followPosition: true,
  marker: [],
  onMarkerPressed: jest.fn()
};

it('renders without crashing', () => {
  const tree = shallow(<Map {...props} />);
  expect(tree).toBeTruthy();
});

it('matches snapshot', () => {
  const tree = shallow(<Map {...props} />);
  expect(tree).toMatchSnapshot();
});

it('should not render without position', () => {
  const tree = shallow(<Map {...props} />);
  tree.setProps({
    position: null
  });
  expect(tree.getElement()).toBeNull();
});

it('handles position updates correctly', () => {
  const tree = shallow(<Map {...props} />);
  tree.instance().map = {
    animateToCoordinate: jest.fn()
  };
  const positionUpdate = {
    position: {
      coords: {
        latitude: 55,
        longitude: 15
      }
    }
  };
  tree.setProps(positionUpdate);
  tree.update();
  expect(tree.instance().map.animateToCoordinate).toHaveBeenCalledWith(
    positionUpdate.position.coords
  );
});

it('only animates map if followPosition is turned on', () => {
  const tree = shallow(<Map {...props} />);
  tree.instance().map = {
    animateToCoordinate: jest.fn()
  };
  tree.setProps({
    followPosition: false
  });
  tree.update();
  expect(tree.instance().map.animateToCoordinate).not.toHaveBeenCalled();
  tree.setProps({
    followPosition: true
  });
  tree.update(); // should animate map
  tree.setProps({
    position: {
      coords: {
        latitude: 55,
        longitude: 15
      }
    }
  });
  tree.update(); // should animate map
  tree.setProps({
    followPosition: false
  });
  tree.update();
  tree.setProps({
    position: {
      coords: {
        latitude: 55,
        longitude: 15
      }
    }
  });
  tree.update();
  expect(tree.instance().map.animateToCoordinate).toHaveBeenCalledTimes(2);
});

it('handles marker updates correctly', () => {
  const tree = shallow(<Map {...props} />);
  expect(tree.find('MockMarker')).toHaveLength(0);
  const marker = [
    {
      id: 'markerid',
      coordinate: {
        latitude: 60,
        longitude: 20
      }
    },
    {
      id: 'markerid',
      coordinate: {
        latitude: 70,
        longitude: 25
      }
    }
  ];
  tree.setProps({
    marker
  });
  tree.update();
  expect(tree.find('MockMarker')).toHaveLength(2);
  expect(
    tree
      .find('MockMarker')
      .first()
      .prop('coordinate')
  ).toMatchObject(marker[0].coordinate);
  expect(
    tree
      .find('MockMarker')
      .at(1)
      .prop('coordinate')
  ).toMatchObject(marker[1].coordinate);
});

it('handles route updates correctly', () => {
  const tree = shallow(<Map {...props} />);
  expect(tree.find('MockPolyline')).toHaveLength(0);
  const route = [
    {
      latitude: 60,
      longitude: 20
    },
    {
      latitude: 65,
      longitude: 25
    },
    {
      latitude: 70,
      longitude: 30
    }
  ];
  tree.setProps({
    route
  });
  tree.update();
  expect(tree.find('MockPolyline')).toHaveLength(1);
  expect(tree.find('MockPolyline').prop('coordinates')).toMatchObject(route);
});

it('reacts to marker press events correctly', () => {
  const tree = shallow(<Map {...props} />);
  expect(tree.find('MockMarker')).toHaveLength(0);
  const marker = [
    {
      id: 'markerid',
      coordinate: {
        latitude: 60,
        longitude: 20
      }
    },
    {
      id: 'markerid',
      coordinate: {
        latitude: 70,
        longitude: 25
      }
    }
  ];
  tree.setProps({
    marker
  });
  tree.update();
  tree
    .find('MockMarker')
    .first()
    .simulate('press');
  expect(props.onMarkerPressed).toHaveBeenCalledWith(marker[0]);
});

it('should handle map press events', () => {
  const onMapLongPressed = jest.fn();
  const p = { ...props, onMapLongPressed };
  const tree = shallow(<Map {...p} />);
  tree.find('MockMapView').simulate('longPress');
  expect(onMapLongPressed).toHaveBeenCalledTimes(1);
});
