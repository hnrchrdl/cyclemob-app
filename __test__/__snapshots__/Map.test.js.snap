// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Map
    followPosition={true}
    marker={Array []}
    position={
      Object {
        "coords": Object {
          "latitude": 50,
          "longitude": 50,
        },
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": Array [
        <MockUrlTile
          urlTemplate="https://tile.thunderforest.com/cycle/{z}/{x}/{y}.png?apikey=your_api_key_here"
          zIndex={-1}
        />,
        Array [],
      ],
      "initialRegion": Object {
        "latitude": 50,
        "latitudeDelta": 0.02,
        "longitude": 50,
        "longitudeDelta": 0.011244377811094454,
      },
      "mapType": "None",
      "onLongPress": [Function],
      "showsUserLocation": true,
      "style": Object {
        "bottom": 0,
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
      },
    },
    "ref": [Function],
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "urlTemplate": "https://tile.thunderforest.com/cycle/{z}/{x}/{y}.png?apikey=your_api_key_here",
          "zIndex": -1,
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": Array [
          <MockUrlTile
            urlTemplate="https://tile.thunderforest.com/cycle/{z}/{x}/{y}.png?apikey=your_api_key_here"
            zIndex={-1}
          />,
          Array [],
        ],
        "initialRegion": Object {
          "latitude": 50,
          "latitudeDelta": 0.02,
          "longitude": 50,
          "longitudeDelta": 0.011244377811094454,
        },
        "mapType": "None",
        "onLongPress": [Function],
        "showsUserLocation": true,
        "style": Object {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        },
      },
      "ref": [Function],
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "urlTemplate": "https://tile.thunderforest.com/cycle/{z}/{x}/{y}.png?apikey=your_api_key_here",
            "zIndex": -1,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
